Internal traffic policy

gentoo-root opened the pull request:
Internal traffic policy = Local means that internal (East-West)
connections to the service should be handled by node-local backends
only. When external and internal traffic policies are different, the
existing implementation is extended to support filtering out non-local
backends for the internal scope. When both policies are both Local, only
the external scope is used (the same already happens when both policies
are Cluster), and it contains local backends only.

Fixes: #16071

```release-note
Support Kubernetes v1.21 new field internalTrafficPolicy=Local.
```

gentoo-root committed:
bpf: Rename TrafficPolicy to ExtTrafficPolicy

K8S services have two traffic policies: external and internal, each of
which can be one of Local or Cluster. So far, Cilium only supported the
externalTrafficPolicy, and it has been called just "TrafficPolicy" all
over the code. As a preparation for internalTrafficPolicy support,
rename existing occurrences of "TrafficPolicy" to "ExtTrafficPolicy".

For backward compatibility, preserve the old field names in API
structures, however, mark them as deprecated and add the new fields
specific for the external traffic policy.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Add support for internalTrafficPolicy=Local

Internal traffic policy = Local means that internal (East-West)
connections to the service should be handled by node-local backends
only. When external and internal traffic policies are different, the
existing implementation is extended to support filtering out non-local
backends for the internal scope. When both policies are both Local, only
the external scope is used (the same already happens when both policies
are Cluster), and it contains local backends only.

 Int policy | Ext policy | Behavior
------------+------------+----------------------------------------------
 Cluster    | Cluster    | External scope contains all backends.
            |            | Internal scope is empty and not queried.
            |            | North-South traffic uses external scope.
            |            | East-West traffic uses external scope.
------------+------------+----------------------------------------------
 Cluster    | Local      | External scope contains node-local backends.
            |            | Internal scope contains all backends.
            |            | North-South traffic uses external scope.
            |            | East-West traffic uses internal scope.
------------+------------+----------------------------------------------
 Local      | Cluster    | External scope contains all backends.
            |            | Internal scope contains node-local backends.
            |            | North-South traffic uses external scope.
            |            | East-West traffic uses internal scope.
------------+------------+----------------------------------------------
 Local      | Local      | External scope contains node-local backends.
            |            | Internal scope is empty and not queried.
            |            | North-South traffic uses external scope.
            |            | East-West traffic uses external scope.

Note that ClusterIP is a special case, since it doesn't have any
externalTrafficPolicy. ClusterIP services always have one scope
(external), which contains either backends according to the
internalTrafficPolicy. It allows to keep the optimization added in
commit a2086bcc5b3c ("bpf: only compile-in scope-switch for svc lookup
on nodeport").

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gandro reviewed (commented):
I'm very excited to see this land! :tada: 

The renaming of `trafficPolicy` fields in the `api/` subfolder (both Hubble and OpenAPI), as well as the monitor API might break external JSON parsers trying to read the field. For example, Hubble can export those events into a JSON file which are then read by external systems.

I don't think we should rename it without warning/deprecation. If we want to rename it, I would suggest adding a new field called externalTrafficPolicy, populate both the new and old one, and deprecate the old one. Later, we can then check if it is safe to remove the old field.

gentoo-root committed:
bpf: Add an integration test for external and internal traffic policy

The new test checks that the right subsets of backends are assigned to
services with different external and internal traffic policies.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Extend the loadbalancer flags test to check internal scope

The previous commit added support for internalTrafficPolicy=Local, which
is indicated by the serviceFlagIntLocalScope load balancer flag. Extend
the existing test to check all possible combinations of external and
internal traffic policies. Also test conversion to string for new flags.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

borkmann commented on bpf/lib/common.h:
```diff
@@ -726,6 +726,7 @@ enum {
 	SVC_FLAG_LOADBALANCER = (1 << 5),  /* LoadBalancer service */
 	SVC_FLAG_ROUTABLE     = (1 << 6),  /* Not a surrogate/ClusterIP entry */
 	SVC_FLAG_SOURCE_RANGE = (1 << 7),  /* Check LoadBalancer source range */
+	SVC_FLAG_INT_LOCAL_SCOPE = (1 << 12), /* internalTrafficPolicy=Local */
```
nit: should this not go into flags2 below given here we're already using all 8 bits? (See the SVC_FLAG_LOOPBACK handling.)

borkmann replied:
nit: Could you extend the commit message on how the result looks like with a simple service in k8s with combinations of for example:

- internalTrafficPolicy={Local,Cluster}
- externalTrafficPolicy={Local,Cluster}
- internalTrafficPolicy={Local,Cluster} and externalTrafficPolicy={Local,Cluster}

gentoo-root replied:
Done.

> Could you extend the commit message

Maybe it's better to put the table from the commit message somewhere in the documentation? The commit message is harder to find, and it can't be amended in the future. What's the best place in the docs to put this sort of details?

borkmann replied:
In terms of documentation we do have a section in here: https://docs.cilium.io/en/latest/network/kubernetes/kubeproxy-free/#client-source-ip-preservation (Documentation/network/kubernetes/kubeproxy-free.rst), but this is mainly user facing. I think we should definitely extend the documentation on this as well that we support this too. In terms of implementation details iow how the svc/bpf map gets populated, we don't have a concrete document hence my thinking that this could perhaps go as part of the commit message. Other option could be to add a paragraph into the doc explaining the behavior with svc list examples and what these entries do.

julianwiedmann replied:
> Maybe it's better to put the table from the commit message somewhere in the documentation?

This is about the table in patch 2? I feels that level of detail actually works well in the commit description - it reflects the current implementation, and is the natural point where developers will go when trying to understand it.

julianwiedmann commented on pkg/loadbalancer/loadbalancer_test.go:
```diff
@@ -331,12 +413,12 @@ func TestServiceFlags_String(t *testing.T) {
 		{
 			name: "Test-4",
-			s:    serviceFlagExternalIPs | serviceFlagLocalScope | serviceFlagRoutable,
+			s:    serviceFlagExternalIPs | serviceFlagExtLocalScope | serviceFlagRoutable,
```
nit - maybe also extend this with some additional tests for the various scope combinations?

gentoo-root replied:
Done.

julianwiedmann commented on cilium/cmd/service_update.go:
```diff
@@ -50,7 +50,7 @@ func init() {
 	serviceUpdateCmd.Flags().BoolVarP(&k8sHostPort, "k8s-host-port", "", false, "Set service as a k8s HostPort")
 	serviceUpdateCmd.Flags().BoolVarP(&localRedirect, "local-redirect", "", false, "Set service as Local Redirect")
-	serviceUpdateCmd.Flags().StringVarP(&k8sTrafficPolicy, "k8s-traffic-policy", "", "Cluster", "Set service with k8s externalTrafficPolicy as {Local,Cluster}")
+	serviceUpdateCmd.Flags().StringVarP(&k8sExtTrafficPolicy, "k8s-ext-traffic-policy", "", "Cluster", "Set service with k8s externalTrafficPolicy as {Local,Cluster}")
```
Do we need to update/extend the following? https://github.com/cilium/cilium/blob/f59df8580b559aa122c3f0ff9d322360dbff7ef9/test/helpers/kubectl.go#L4165

gentoo-root replied:
Yes, good catch.

Seems that this code is never called with trafficPolicy = Local. Also it seems like the only caller is https://github.com/cilium/cilium/blob/f59df8580b559aa122c3f0ff9d322360dbff7ef9/test/k8s/datapath_configuration.go#L333, and I wonder, if it's a test, why no tests have failed?

BTW, do we want to keep `--k8s-traffic-policy` for backward compatibility?

julianwiedmann replied:
> and I wonder, if it's a test, why no tests have failed?

The e2e tests need an additional trigger. I just kicked them off. Think the error would have looked similar as [this one](https://github.com/cilium/cilium/issues/21392).

> BTW, do we want to keep --k8s-traffic-policy for backward compatibility?

Yeah, was wondering about the same. A quick search didn't find any external users/docs/publications etc (except our cmd reference), so I would question whether the flag has widespread usage. Let's rip it out, and add it back if anyone complains? :)

julianwiedmann reviewed (changes requested):
Very cool, thank you Maxim! That's a good chunk of work :).

Looks good from what I can tell. Just some tiny adjustments needed for the e2e tests?

gandro reviewed (approved):
Hubble changes look good to me now. Thanks!

michi-covalent reviewed (approved):
hubble-related change LGTM

julianwiedmann reviewed (approved):
Thanks!

nebril commented on cilium/cmd/service_update.go:
```diff
@@ -50,8 +51,9 @@ func init() {
 	serviceUpdateCmd.Flags().BoolVarP(&k8sLoadBalancer, "k8s-load-balancer", "", false, "Set service as a k8s LoadBalancer")
 	serviceUpdateCmd.Flags().BoolVarP(&k8sHostPort, "k8s-host-port", "", false, "Set service as a k8s HostPort")
 	serviceUpdateCmd.Flags().BoolVarP(&localRedirect, "local-redirect", "", false, "Set service as Local Redirect")
-	serviceUpdateCmd.Flags().StringVarP(&k8sTrafficPolicy, "k8s-traffic-policy", "", "Cluster", "Set service with k8s externalTrafficPolicy as {Local,Cluster}")
```
Shouldn't this option be deprecated for one release before being deleted? I don't see it in a list of deprecated options in 1.12.

gentoo-root replied:
@julianwiedmann said above:
> > BTW, do we want to keep --k8s-traffic-policy for backward compatibility?
>
> Yeah, was wondering about the same. A quick search didn't find any external users/docs/publications etc (except our cmd reference), so I would question whether the flag has widespread usage. Let's rip it out, and add it back if anyone complains? :)

If the rule is to keep it as deprecated for one release, I can bring it back.

borkmann replied:
Given this the `service update` cmdline is mainly used for local testing, we don't need compat imho.

joestringer replied:
Should we mark these flags as hidden if they're not for users?

nebril reviewed (changes requested):
Most changes look good to me, one question about deprecating old option left inline.

borkmann commented on Documentation/network/kubernetes/kubeproxy-free.rst:
```diff
@@ -324,6 +324,35 @@ be lost on its path to the service endpoint.
++----------+----------+-------------------------+-----------------------+
+| Internal | External | for North-South traffic | for East-West traffic |
++==========+==========+=========================+=======================+
+| Cluster  | Cluster  | All                     | All                   |
```
just a tiny nit: could you change `All` into `All (default)` so it is obvious to readers what is the default behavior?

brb commented on bpf/lib/lb.h:
```diff
@@ -283,15 +283,15 @@ static __always_inline bool lb_skip_l4_dnat(void)
+bool lb4_svc_is_two_scopes(const struct lb4_service *svc)
 {
-	return svc->flags & SVC_FLAG_EXT_LOCAL_SCOPE;
+	return (svc->flags & SVC_FLAG_EXT_LOCAL_SCOPE) != (svc->flags2 & SVC_FLAG_INT_LOCAL_SCOPE);
```
I'm a bit puzzled by this check and the function name. Is this function supposed to check whether any scope is set (external, local, or both)? From the check it looks that's the case. However, the name reads as if it checks for both being set. Maybe then `lb4_svc_has_any_scope()`?

borkmann replied:
I think it's okay as is (maybe comment could help here). If both scopes are Local, then we don't use the two lookup scopes since the set of backends for both cases is the same.

gentoo-root replied:
This line is buggy, by the way. The two bits are at different positions, I have to convert both sides of `!=` to bool first.

brb commented on bpf/lib/lb.h:
```diff
@@ -1225,7 +1225,7 @@ struct lb4_service *lb4_lookup_service(struct lb4_key *key,
 	key->backend_slot = 0;
```
In the case of socket-LB being disabled (e.g., when `--kube-proxy-replacement=disabled`), the bpf_lxc does the E/W LB for Pods running in non-host netns.  In this case, the `lb{4,6}_lookup_service()` is called in bpf_lxc.c for egressing pkts.

I'm wondering how the internal traffic policy is going to be handled for such a case? Don't we need to switch the scope above to `LB_LOOKUP_SCOPE_INT`?

borkmann replied:
Looks like right now we only do this in bpf_lxc for kpr enabled: `lb{4,6}_lookup_service(&key, is_defined(ENABLE_NODEPORT),  ...)`. Agree that we should call it unconditionally with `lb{4,6}_lookup_service(&key, true,  ...)` and test whether for ClusterIP it matches also kube-proxy behavior.

gentoo-root replied:
> Looks like right now we only do this in bpf_lxc **for kpr enabled**: lb{4,6}_lookup_service(&key, is_defined(ENABLE_NODEPORT), ...)

Only for partial kpr (strict kpr disables this whole block with `#ifdef ENABLE_PER_PACKET_LB`).

I'm a bit puzzled, why we have this `is_defined(ENABLE_NODEPORT)` there. I.e. why do we pick the internal scope if ENABLE_NODEPORT is enabled, and the external scope otherwise?

Also, I couldn't actually get into this flow in bpf_lxc, where ENABLE_PER_PACKET_LB is true, and ENABLE_NODEPORT is false (kpr is disabled). All connections to ClusterIP go through bpf_sock, and I don't see an option to disable it.

> test whether for ClusterIP it matches also kube-proxy behavior.

The behavior of kube-proxy (on k8s 1.25) doesn't look sane to me. Connections to a NodePort at the internal IP of the local node doesn't obey any policy (all backends are returned regardless of the external and internal traffic policy), and other kinds of connections to a NodePort (i.e. to external IPs, or to the internal IP of a remote node) obey the external traffic policy.

gentoo-root replied:
> I'm a bit puzzled, why we have this is_defined(ENABLE_NODEPORT) there.

Figured out this one, it's an optimization (https://github.com/cilium/cilium/commit/a2086bcc5b3c6071db4078c95fc2dc0d6093de36), which relies on the fact that ClusterIP services don't have two scopes, but I changed it in this commit, so I guess I'll have to revert it.

brb reviewed (commented):
Amazing work! Two questions/clarifications needed (it's been a while since I touched that code).

joestringer commented on api/v1/flow/flow.proto:
```diff
@@ -663,9 +663,11 @@ message ServiceUpsertNotification {
     repeated ServiceUpsertNotificationAddr backend_addresses = 3;
     string type = 4;
-    string traffic_policy = 5;
+    string traffic_policy = 5 [deprecated = true];
```
Do we have an idea when we plan to remove this? I realize this could break Hubble flow handling so we may want to wait a few releases. But it would be nice to add a comment like `// TODO: Remove after v1.16` or similar.

kaworu replied:
FWIW we usually do not plan to remove deprecated fields for compatibility reasons (e.g. `reply` or `drop_reason` in the Flow message).

joestringer reviewed (approved):
Docs structure looks good, I didn't review the content as I'm not that familiar with the details of internalTrafficPolicy vs externalTrafficPolicy. Others looked at BPF side already so :+1: 

One minor comment on deprecation + adding a comment to highlight when we can remove the deprecated field. Might need @cilium/sig-hubble input to decide how long to keep the old field around.

kaworu reviewed (approved):
Hubble changes LGTM!

gentoo-root committed:
k8s: Fix node ExternalIP resetting when IPv6 is unused

When there is no IPv6, SetK8sExternalIPv6 is called with nil as an
argument, and it passes it further to SetNodeExternalIP. However, the
latter doesn't make a distinction between IPv4 and IPv6, and if nil
comes in, it just removes all ExternalIPs it knows about, effectively
removing the IPv4 ExternalIP.

Fix this mistake by skipping the call to SetK8sExternalIPv6 if it is
going to remove the address.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
pkg/datapath: Add node ExternalIPs to ipcache

Socket load balancer uses the ipcache to lookup service endpoints. For
NodePort services, ipcache contains an entry with the InternalIP of the
local node, and a wildcard with the 0.0.0.0 IP. The latter is used to
match on InternalIPs and ExternalIPs of remote nodes.

However, the ExternalIP of the local node is not covered, which means
that socket load balancer fails to find a backend, and Cilium falls back
to per-packet load balancing. Another side effect is that the
externalTrafficPolicy will be applied to such internal traffic.

Fix it by adding the ExternalIP of the local node to the ipcache along
with the InternalIP, allowing socket load balancer to match on it.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Block service connections if no backend is available

Similar to commit 183501124869 ("bpf: drop SVC traffic if no backend is
available"), make socket load balancer block connections to the services
that have no backends (either no backends at all, or no local backends
when the internalTrafficPolicy is set to Local).

Currently socket load balancer allows all connections, even if it was a
service endpoint without available backends. In such case, packets fall
back to per-packet load balancing, instead of being dropped, which has
an undesired effect: if internalTrafficPolicy is Local and there are no
local backends, per-packet load balancer will retry the lookup in the
external scope, and if externalTrafficPolicy is Cluster, the client will
end up connecting to some remote backend, contrary to the internal
traffic policy.

Fix it by making syscalls return an error when there are no local
backends available.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Sync linux/bpf.h to upstream

The new header is needed for the definition of bpf_set_retval.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Return -EADDRINUSE from post_bind where possible

-EADDRINUSE is the valid error code to indicate that the requested
address is already taken. Use bpf_set_retval to return this error code
to the userspace (instead of the default -EPERM) on kernels where this
helper is available.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Optimize lb{4,6}_svc_is_two_scopes check

Add a flags that indicates that two lookup scopes (external and
internal) are present, in order to reduce the number of arithmetic and
bit operations per lookup.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
k8s-upstream: Reenable the internalTrafficPolicy test

After the internalTrafficPolicy functionality is implemented, reenable
the corresponding k8s upstream test. Partially addresses #17796.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -452,10 +454,19 @@ __sock4_health_fwd(struct bpf_sock_addr *ctx __maybe_unused)
+	if (err == -EHOSTUNREACH || err == -ENOMEM) {
+#ifdef HAVE_SET_RETVAL
+		bpf_set_retval(err);
+#endif
```
tiny nit, but not blocking: I'd make a small static inline helper like `set_retval(err)` and just hide the ifdef there, so it doesn't need to be on all callsites.

gentoo-root replied:
Done.

pchaigno commented on pkg/k8s/init.go:
```diff
@@ -182,7 +182,10 @@ func WaitForNodeInformation(ctx context.Context, k8sGetter k8sGetter) error {
-		node.SetK8sExternalIPv6(n.GetExternalIP(true))
+		// Avoid removing the IPv4 if there is no IPv6.
+		if extIP6 := n.GetExternalIP(true); extIP6 != nil {
+			node.SetK8sExternalIPv6(extIP6)
```
Don't we need the same fix for when IPv4 is disabled?

pchaigno replied:
I think as Julian said, given the user impact, we probably want to backport this commit to v1.12. Since it wasn't made in a separate pull request, I think you'll need to backport it manually. 

See https://docs.cilium.io/en/latest/contributing/release/backports/#preparation for the documentation. I'm happy to help of course.

gentoo-root replied:
>Don't we need the same fix for when IPv4 is disabled?

No, because the first SetK8sExternalIPv4 will delete all IPs, and the second SetK8sExternalIPv6 will add the IPv6.

pchaigno replied:
Oh, right :facepalm:

pchaigno commented on bpf/bpf_sock.c:
```diff
@@ -452,10 +454,17 @@ __sock4_health_fwd(struct bpf_sock_addr *ctx __maybe_unused)
+	err = __sock4_xlate_fwd(ctx, ctx, false);
+	if (err == -EHOSTUNREACH || err == -ENOMEM) {
+		try_set_retval(err);
+		return SYS_REJECT;
```
Couldn't we also do that for code that returns `SYS_REJECT` and existed before this commit? I see you did that for some BPF programs in a subsequent commit, but it seems it's still missing for `__sock{4,6}_health_fwd` and `cgroup/bind{4,6}`.

gentoo-root replied:
Good idea, I'll just need to figure out the right error codes for those cases.

gentoo-root replied:
https://github.com/cilium/cilium/pull/22965

pchaigno reviewed (commented):
I'm late to the party, but nice work! The commit history is neat and makes things easy to review.

A couple comments below on things we may want to follow up on, if only by opening issues to track them.
